Code to remove qt3 versus qt4/qt5 conflicts when building cbtf-argonavis-gui

jgalarowicz opened the pull request:
@scheibelp This pull request contains code to remove qt3 versus qt4/qt5 conflicts when building cbtf-argonavis-gui and solve issue #7946.    It creates an openspeedshop-utils package that does not depend on qt3 and can be used to resolve the needs of cbtf-argonavis-gui.    Changes for creating a release versus develop build path are also included.

adamjstewart commented on var/spack/repos/builtin/packages/cbtf-argonavis-gui/package.py:
```diff
@@ -49,20 +49,55 @@ class CbtfArgonavisGui(QMakePackage):
+    # To specify ^elfutils@0.170 on the command line spack 
+    # apparently needs/wants this dependency explicity here
+    # even though it is referenced downstream
+    depends_on("elf", type="link")
```
Where exactly downstream? As mentioned [here](http://spack.readthedocs.io/en/latest/known_issues.html#default-variants-are-not-taken-into-account-during-concretization), there are known issues with this. Even then, it isn't a good idea to hack the package to work around them.

adamjstewart commented on var/spack/repos/builtin/packages/cbtf-krell/package.py:
```diff
@@ -83,29 +89,53 @@ class CbtfKrell(CMakePackage):
-    depends_on("binutils")
+    # For binutils
+    depends_on("binutils", when='@develop')
+    depends_on("binutils@2.29.1", when='@1.9.1.0')
```
Does this directly rely on binutils? See http://spack.readthedocs.io/en/latest/getting_started.html#binutils

jgalarowicz replied:
Yes, we use the libiberty library and libbfd to resolve symbols.

jgalarowicz replied:
@adamjstewart I think I missed the comment about all the duplication.   Is this referring to the same package having two depends_on statements?   My co-worked wants a release path and development path through the spack build process.   So, we created a release branch 1.9.1.0 for the cbtf packages and 2.3.1.3 for the openspeedshop packages.   Is there a better way to do what we are attempting?
Thanks
Jim G

adamjstewart commented on var/spack/repos/builtin/packages/openspeedshop-utils/package.py:
```diff
@@ -0,0 +1,343 @@
+       IA32, EM64T, AMD64, PPC, ARM, Power8, Intel Phi, Blue Gene and
+       Cray platforms.  OpenSpeedShop development is hosted by the Krell
+       Institute. The infrastructure and base components of OpenSpeedShop
+       are released as open source code primarily under LGPL.
```
This description doesn't really tell me anything about what this package is for.

jgalarowicz replied:
Ok.  I need to explain what openspeedshop-utils is.  I will update.

adamjstewart commented on var/spack/repos/builtin/packages/openspeedshop-utils/package.py:
```diff
@@ -0,0 +1,343 @@
+
+        # Common settings to both offline and cbtf versions
+        # of OpenSpeedShop
+        run_env.prepend_path('PATH', self.prefix.bin)
```
This shouldn't be necessary, it is already present in Spack by default.

jgalarowicz replied:
Ok.  I will take that out.

adamjstewart commented on var/spack/repos/builtin/packages/openspeedshop-utils/package.py:
```diff
@@ -0,0 +1,343 @@
+        run_env.prepend_path('PATH', self.spec['mrnet'].prefix.bin)
+        run_env.prepend_path('PATH', self.spec['cbtf-krell'].prefix.bin)
+        run_env.prepend_path('PATH', self.spec['cbtf-krell'].prefix.sbin)
+        run_env.prepend_path('PATH', self.spec['python'].prefix.bin)
```
Most Spack packages don't add dependencies to their module files in order to prevent things from being polluted.

jgalarowicz replied:
I was having problems executing openspeedshop, so I added these.   I added the python one recently because without it I was getting an error related to python.  I either had to manually load the python module or add this to the module file.

jgalarowicz replied:
Because when I user loads a openspeedshop module outside of spack, all dependencies are taken care of.  Having to load another module hinders usuage.   Plus the user wouldn't know what module to load.

jgalarowicz replied:
when a user

jgalarowicz commented:
I have no idea what flake wants to solve this issue at line 326 in openspeedshop-utils:
        # Find openspeedshop library path
        oss_libdir = find_libraries('libopenss-framework',
                           root=self.spec['openspeedshop-utils'].prefix,
                           shared=True, recursive=True)

jgalarowicz commented:
```
oss_libdir = find_libraries('libopenss-framework',
                   root=self.spec['openspeedshop-utils'].prefix,
                   shared=True, recursive=True)
```

adamjstewart commented:
Seems like it worked this time. For future reference, if you run `pip install flake8` and then run `spack flake8` you can instantly check whether or not your changes worked. Also, I know @alalazo likes autopep8 for automatically resolving flake8 warnings.

scheibelp commented:
I'll be able to look at this tomorrow (5/11)

jgalarowicz commented:
@scheibelp Thanks!

scheibelp commented on var/spack/repos/builtin/packages/openspeedshop-utils/package.py:
```diff
@@ -0,0 +1,345 @@
+        instrumentor_setting = "cbtf"
+
+        if spec.satisfies('+runtime'):
+            # Appends base options to cmake_args
```
Not sure what this comment adds (e.g. it's the same in both branches of the if statement)

jgalarowicz replied:
I will remove.

scheibelp commented on var/spack/repos/builtin/packages/openspeedshop-utils/package.py:
```diff
@@ -0,0 +1,345 @@
+                 '-DCBTF_KRELL_DIR=%s' % spec['cbtf-krell'].prefix,
+                 '-DMRNET_DIR=%s' % spec['mrnet'].prefix])
+
+            cmake_args.extend(['-DBUILD_QT3_GUI=FALSE'])
```
Is the idea that `+runtime` builds a private instance of qt 3? Or is it that one will always be building qt 4/5 when `+runtime` is set?

jgalarowicz commented on var/spack/repos/builtin/packages/cbtf-argonavis-gui/package.py:
```diff
@@ -49,20 +49,55 @@ class CbtfArgonavisGui(QMakePackage):
-    depends_on("cbtf")
-    depends_on("cbtf-krell")
-    depends_on("cbtf-argonavis")
+    version('develop', branch='master',
```
@scheibelp We left those in because without that in we can't remove those old 1.3.0 packages.  Spack gets an abort when we try to remove the old packages.  So, we left those in even though they point to the same "master" version.

scheibelp replied:
Ah got it - we should make that easier to do

scheibelp replied:
@jgalarowicz I just tried uninstalling a package after removing the version and that appeared to work. Could you post the error message you get when you try this?

scheibelp commented on var/spack/repos/builtin/packages/cbtf-argonavis-gui/package.py:
```diff
@@ -49,20 +49,55 @@ class CbtfArgonavisGui(QMakePackage):
+
+    # For MRNet
+    depends_on("mrnet@5.0.1-3:+lwthreads", when='@develop')
+    depends_on("mrnet@5.0.1-3:+lwthreads", when='@1.3.0.0')
```
To be clear: when you have a `:` in this case it means "version `5.0.1-3` or greater", and also note that the `:` is present on both of the `mrnet` `depends_on` lines (so it sounds like you would want to remove it from the 2nd line).

jgalarowicz replied:
Ok, that was a mistake on my part, I will fix that - take off the colon.

jgalarowicz commented on var/spack/repos/builtin/packages/cbtf-argonavis/package.py:
```diff
@@ -65,17 +71,52 @@ class CbtfArgonavis(CMakePackage):
+    depends_on("boost@1.50.0:", when='@develop')
+    depends_on("boost@1.66.0", when='@1.9.1.0')
+
+    # For MRNet
```
Ok, I will change cbtf-argonavis package to use the suggested change.

jgalarowicz commented on var/spack/repos/builtin/packages/cbtf-argonavis/package.py:
```diff
@@ -109,8 +150,8 @@ def cmake_args(self):
+        cupti_path = self.spec['cuda'].prefix + '/extras/CUPTI/lib64'
         if os.environ.get('LD_LIBRARY_PATH'):
-            cupti_path = self.spec['cuda'].prefix + '/extras/CUPTI/lib64'
             os.environ['LD_LIBRARY_PATH'] += cupti_path
```
I will look into this and make the change to prepend_path.

jgalarowicz commented on var/spack/repos/builtin/packages/cbtf-lanl/package.py:
```diff
@@ -48,26 +48,57 @@ class CbtfLanl(CMakePackage):
+    version('1.9.1.0', branch='1.9.1.0',
+            git='https://github.com/OpenSpeedShop/cbtf-lanl.git')
+
     version('1.9.1', branch='master',
```
This is the issue with being able to remove the old packages that we created w/o having a corresponding specific branch that matches the release numbers (1.3.0 and 1.9.1).

jgalarowicz commented on var/spack/repos/builtin/packages/openspeedshop/package.py:
```diff
@@ -112,48 +110,77 @@ class Openspeedshop(CMakePackage):
-    depends_on("binutils", type='build')
+
+    # For binutils
+    depends_on("binutils", when='@develop')
```
This is a mistake on my part.  I will fix.

jgalarowicz commented:
@scheibelp @adamjstewart I think I addressed the issues that were brought up and submitted the changes.    I left the 1.9.1, 1.3.0, and 2.3.1 version lines in, so those versions can be removed.  But, I can take those out if you want me to.

scheibelp commented on var/spack/repos/builtin/packages/openspeedshop/package.py:
```diff
@@ -112,48 +110,77 @@ class Openspeedshop(CMakePackage):
+    depends_on("mrnet@5.0.1-3:+cti", when='@develop+cti')
+    depends_on("mrnet@5.0.1-3:+lwthreads", when='@develop')
+
+    depends_on("mrnet@5.0.1-3:+cti", when='@2.3.1.3+cti')
```
@jgalarowicz did you mean `mrnet@5.0.1-3+cti` vs. `mrnet@5.0.1-3:+cti`? i.e. did you want to be exact about the version? If not, these two lines can be merged with the above two lines.

jgalarowicz replied:
@scheibelp Yes, we wanted the exact version for the release type path through the O|SS packages.    The release would be more stable based on using packages we know to work with the rest of the tool.

jgalarowicz commented:
@scheibelp Hi Peter - is there anything else you want done for this PR?   Thanks - Jim G

scheibelp commented:
Hi @jgalarowicz I was curious about the error message you got when you removed the old versions (e.g. `1.3.0` for `cbtf-argonavis-gui`) as I was having trouble reproducing that issue: https://github.com/spack/spack/pull/8066#discussion_r188386709

jgalarowicz commented:
@scheibelp Hi Peter - I posted what I'm seeing when trying to uninstall.   In the place where you originally asked the question.   Sorry - I missed that question when I was responding to issues.

scheibelp commented:
Thanks for posting the output! That looks like a bug. I'll merge this and file an issue on it.
